Maintainers blogpost (#2684)
nf-core/website: sites/main-site/src/content/blog/2024/maintainers-minutes-2024-07-26.mdx

---
title: 'Maintainers Minutes: August 2024'
subtitle: 'Keeping you informed of the latest maintainers discussions'
pubDate: 2024-08-14T12:00:00+02:00

> headerImage: '/assets/images/blog/maintainers-minutes-2024-07-26/maintainers_duck.png'

Reviewer: This is huge when rendering on the website, can we limit the size of the header @mashehu ?

Reviewer 2: Done. there is now a headerImageDim option to control that.

headerImageAlt: "Cartoon yellow rubber duck with nf-core logo badge on it's body with the nf-core logo."
embedHeaderImage: true
authors:
    - 'jfy133'
    - 'FriederikeHanssen'
label:
    - 'maintainers'
---

The 'Maintainers Minutes' aims to further give insight into the workings of the [nf-core maintainers team](/governance#maintainers)
by providing brief summaries of the monthly team meetings.

> ## Review of action points from last time
> 
> We got about half way through our action points from last time.
> 
> Slack channels were created for the new working groups: - #wg-meta-pipelines - #wg-modules-update - #wg-test-data-task-force
> 
> Please join if you want to be involved in one of those.

Reviewer: ```suggestion
```

I think we already covered this in the previous blog post, and reviewing last meetings points isn't so interesting to readers (I think anyway)

## More (or less) testing

> Júlia encountered several failing tests on nf-core/crisprseq with conda and singularity. This was caused by tests only running
> on docker. Moving forward we agreed that at least some tests should also be run
> with `singularity` and `conda`.
> 
> We discussed several ideas. On the one hand we'd like to make sure everything works at all times, on the other hand running tons of tests
> makes development slow, costs money, and is not great for the planet 🌳. As a compromise, we agreed on the following: - Singularity & conda tests are run on PR to master - Make singularity required - Conda optional opt-out - Add `workflow dispatch` button for easier manual testing

Reviewer: ```suggestion
Júlia encountered several failing tests on nf-core/crisprseq with conda and singularity (currently, our CI tests only run with docker).
Moving forward we agreed that at least some tests should also be run with `singularity` and `conda` at some point during the development process but before release.

We discussed several ideas. 
On the one hand we'd like to make sure everything works at all times, on the other hand running tons of essentially duplicate tests makes development slow, costs money, and is not great for the planet 🌳.
As a compromise, we agreed on the following: 

- Singularity & conda tests are run on PR to master 
- Make singularity required 
- Conda optional opt-out 
- Add `workflow dispatch` button for easier manual testing during earlier stages of development
```

## Nf-test best practices

> Nicolas brought up that snapshots should be readable to quickly determine that they do what they should. Since the order of snapshots in the `.snap` file
> does not correspond to the order they are defined in the test file, having one snapshot per file in the same test
> scrambles the entries all over the place.
> 
> Nicolas will make updates to the guidelines that all assertions should go into a single snapshot per test.

Reviewer: ```suggestion
Nicolas brought up that nf-test snapshots (i.e., files that record e.g. md5sums or presence of strings) should be readable to quickly determine that they include everything they should.
He had noticed that some people in module tests had started using a snapshot for every single file (rather than all in one), seemingly to be able to 'label' each snapshot.
However, since the order of each 'snapshot' recorded in the `.snap` file does not necessarily correspond to the order they are defined in the test file, having one snapshot _per file_ in the same test scrambles the entries all over the place, e.g. out of order from the output channels as defined in the module.

Nicolas will make updates to the guidelines that all assertions should go into a single snapshot per test for modules.
```

## Bulk updates of modules

> Júlia is coordinating the bulk update of modules in https://github.com/nf-core/modules/issues/5828 (join #wg-modules-update to stay tuned)

Reviewer: ```suggestion
Júlia is coordinating an upcoming bulk update of all modules in https://github.com/nf-core/modules/issues/5828 (join #wg-modules-update to stay tuned)
```

> This includes: - meta.yml accounting for the channel structure (https://github.com/nf-core/modules/pull/5867) - subworkflows - bio.tools identifiers - edam ontologies for all input/output files - removing defaults from conda channel - adding lock files - adding version topic channels - adding lint for stub tests for all modules

Reviewer: ```suggestion
This will include a lot of new exciting things:

  - improved `meta.yml` layout accounting for the channel structure (https://github.com/nf-core/modules/pull/5867) 
  - subworkflows 
  - bio.tools identifiers (for better linked metadata across registries)
  - edam ontologies for all input/output files  (for better standardisation of file types)
  - removing defaults from conda channel (to reduce chances of ToS violations)
  - adding lock files (for better conda dependency version stability)
  - adding version topic channels (for simplified version reporting in pipelines)
  - adding lint for stub tests for all modules (to help us move to having complete 'dry run' functionality in pipelines)
```

### Topic channels for `version`

> In addition, we discussed how to incorporate the topics version channels. Edmund created a [POC](https://github.com/nf-core/nascent/pull/150) showing that
> it is possible to handle topic and non-topic version channels together.
> This also requires a pipeline update by all developers, pinning a recent nextflow version, and adding the `preview` flag.
> It was concluded that this should be a hackathon task.

Reviewer: ```suggestion
In addition, we discussed how to incorporate the version aggregation system using Nextflow's new '[topic](https://www.nextflow.io/docs/latest/channel.html#topic)' channel functionality.
Edmund created a [proof of concept](https://github.com/nf-core/nascent/pull/150) showing that it is possible to handle topic and non-topic version channels together.
This would require a pipeline update by all developers, pinning a recent nextflow version, and adding the `preview` flag.
It was concluded that this should be a hackathon task.
```

## Obligatory argument

> After settling somewhat on how we want to name subworkflows in the previous meeting, we now discussed the meta.yml restructuring for subworkflows.
> The meta.yml currently only includes the channel name. It is now being expanded to match the channel construct, like: "meta, bam, bai" (not just files).
> We also argued over the overall channel names.

Reviewer: ```suggestion
After our recent settling( somewhat) on how we want to name subworkflows in the previous meeting, we now discussed the `meta.yml` restructuring for subworkflows.
The `meta.yml` currently only includes the channel name. 
However we are moving now for it being expanded to match the _entire_ channel construct, like: "meta, bam, bai" (not just files).
However, of course, this let us to start arguing over the overall channel names.

>   There are only two hard things in Computer Science: cache invalidation and naming things.
>
>    -- Phil Karlton

_from [https://martinfowler.com/bliki/TwoHardThings.html](https://martinfowler.com/bliki/TwoHardThings.html)_

```

## Versioning of modules guidelines

> Arthur Gymer suggested that the module guidelines should be versioned.
> Everyone think it's a good idea and we discussed how to best approach this since the guidelines are listed on the website, linting is linked to nf-core/tools versions
> and not everything can/is linted for at the moment.
> 
> Some ideas included linking the docs with the tools version and date stamp the docs, have dev docs version to account for incremental updates. We agreed that linting against the tools dev version may cause
> problems with changes during the developemt.
> One of the major technical challenges we discussed were how to best account for various doc versions on the website with the additional complication the the guidelines page is planned to be split up (which James finds 🤢)
> Matthias will think about a few ways on how this could be tackled.

Reviewer: ```suggestion
Moritz Beber and Arthur Gymer from the nf-core community had both a long time ago and more recently suggested that the module guidelines should be versioned.
Everyone think it's a good idea and we discussed how to best approach this since the guidelines are listed on the website, _but_ linting is linked to nf-core/tools versions, _but_ not everything can/is linted for at the moment.

Some ideas included linking the docs with the tools version and date stamp the docs, have dev docs version to account for incremental updates. 
However, it was brought up that that linting against the tools dev version may cause problems with changes that may occur during the development on the dev branch.
One of the major technical challenges that was flag was to best account for various docs versions on the website, with the additional complication the the guidelines page is planned to be split up into one file per guideline point (which James finds 🤢).
Matthias will think about a few ways on how this could be tackled to minimise the amount of file copies that may occur when archiving each version.
```

## Next time

> Mahesh figured out a cool way to handle args in nf-tests without the `nextflow.config` copying issue: https://nfcore.slack.com/archives/C049MBCEW06/p1721042240473589
> We'll discuss this in detail next time.

Reviewer: ```suggestion
Mahesh has figured out an interesting way to handle args in nf-tests without having to make a separate `nextflow.config` file each time for each test (as brought up here[https://nfcore.slack.com/archives/C049MBCEW06/p1721042240473589](https://nfcore.slack.com/archives/C049MBCEW06/p1721042240473589))
We'll discuss this in detail next time.
```

As a compromise, we agreed on the following:

-   Singularity & conda tests are run on PR to master
-   Make singularity required
-   Conda optional opt-out
-   Add `workflow dispatch` button for easier manual testing during earlier stages of development

## Nf-test best practices

Nicolas brought up that nf-test snapshots (i.e., files that record e.g. md5sums or presence of strings) should be readable to quickly determine that they include everything they should.
He had noticed that some people in module tests had started using a snapshot for every single file (rather than all in one), seemingly to be able to 'label' each snapshot.
However, since the order of each 'snapshot' recorded in the `.snap` file does not necessarily correspond to the order they are defined in the test file, having one snapshot _per file_ in the same test scrambles the entries all over the place, e.g. out of order from the output channels as defined in the module.

Nicolas will make updates to the guidelines that all assertions should go into a single snapshot per test for modules.

## Bulk updates of modules

> Júlia is coordinating an upcoming bulk update of all modules in https://github.com/nf-core/modules/issues/5828 (join #wg-modules-update to stay tuned)

Reviewer: ```suggestion
Júlia is coordinating an upcoming bulk update of all modules in https://github.com/nf-core/modules/issues/5828 (join [#wg-modules-update](https://nfcore.slack.com/archives/C07BDPBPLG3) to stay tuned)
```

This will include a lot of new exciting things:

> -   improved `meta.yml` layout accounting for the channel structure (https://github.com/nf-core/modules/pull/5867)
> -   subworkflows
> -   bio.tools identifiers (for better linked metadata across registries)
> -   edam ontologies for all input/output files (for better standardisation of file types)
> -   removing defaults from conda channel (to reduce chances of ToS violations)
> -   adding lock files (for better conda dependency version stability)
> -   adding version topic channels (for simplified version reporting in pipelines)
> -   adding lint for stub tests for all modules (to help us move to having complete 'dry run' functionality in pipelines)

Reviewer: ```suggestion
-   Improved `meta.yml` layout accounting for the channel structure (https://github.com/nf-core/modules/pull/5867)
-   Subworkflows
-   bio.tools identifiers (for better linked metadata across registries)
-   edam ontologies for all input/output files (for better standardisation of file types)
-   Removing defaults from conda channel (to reduce chances of ToS violations)
-   Adding lock files (for better conda dependency version stability)
-   Adding version topic channels (for simplified version reporting in pipelines)
-   Adding lint for stub tests for all modules (to help us move to having complete 'dry run' functionality in pipelines)
```

### Topic channels for `version`

In addition, we discussed how to incorporate the version aggregation system using Nextflow's new '[topic](https://www.nextflow.io/docs/latest/channel.html#topic)' channel functionality.
Edmund created a [proof of concept](https://github.com/nf-core/nascent/pull/150) showing that it is possible to handle topic and non-topic version channels together.
This would require a pipeline update by all developers, pinning a recent nextflow version, and adding the `preview` flag.
It was concluded that this should be a hackathon task.

## Obligatory argument

> After our recent settling( somewhat) on how we want to name subworkflows in the previous meeting, we now discussed the `meta.yml` restructuring for subworkflows.

Reviewer: ```suggestion
After our recent settling (somewhat, anyway) on how we want to name subworkflows in the previous meeting, we now discussed the `meta.yml` restructuring for subworkflows.
```

The `meta.yml` currently only includes the channel name.
However we are moving now for it being expanded to match the _entire_ channel construct, like: "meta, bam, bai" (not just files).
However, of course, this let us to start arguing over the overall channel names.

> There are only two hard things in Computer Science: cache invalidation and naming things.
>
> -- Phil Karlton

_from [https://martinfowler.com/bliki/TwoHardThings.html](https://martinfowler.com/bliki/TwoHardThings.html)_

## Versioning of modules guidelines

> Moritz Beber and Arthur Gymer from the nf-core community had both a long time ago and more recently suggested that the module guidelines should be versioned.

Reviewer: ```suggestion
Various members of the nf-core community had both a long time ago and more recently suggested that the module guidelines should be versioned.
```

> Júlia is coordinating an upcoming bulk update of all modules in https://github.com/nf-core/modules/issues/5828 (join #wg-modules-update to stay tuned)

Reviewer 2: when we mention names, can we link to their github account. makes it feel less insidery

Author: dione
